src/spec-common/injectHeadless.ts

```diff
@@ -667,14 +668,60 @@ async function patchEtcProfile(params: ResolverParameters, containerProperties:
 	}
 }
 
-async function probeUserEnv(params: { defaultUserEnvProbe: UserEnvProbe; allowSystemConfigChange: boolean; output: Log }, containerProperties: { shell: string; remoteExec: ExecFunction; installFolder?: string; env?: NodeJS.ProcessEnv; shellServer?: ShellServer; launchRootShellServer?: (() => Promise<ShellServer>); user?: string }, config?: CommonMergedDevContainerConfig) {
-	const env = await runUserEnvProbe(params, containerProperties, config, 'cat /proc/self/environ', '\0');
-	if (env) {
+async function probeUserEnv(params: { defaultUserEnvProbe: UserEnvProbe; allowSystemConfigChange: boolean; output: Log; containerSessionDataFolder?: string }, containerProperties: { shell: string; remoteExec: ExecFunction; installFolder?: string; env?: NodeJS.ProcessEnv; shellServer?: ShellServer; launchRootShellServer?: (() => Promise<ShellServer>); user?: string }, config?: CommonMergedDevContainerConfig) {
+	let env = await readUserEnvFromCache(params.containerSessionDataFolder, containerProperties.shellServer);
+	if (env && Object.keys(env).length) {
 		return env;
 	}
-	params.output.write('userEnvProbe: falling back to printenv');
-	const env2 = await runUserEnvProbe(params, containerProperties, config, 'printenv', '\n');
-	return env2 || {};
+
+	params.output.write('userEnvProbe: not found in cache');
+	env = await runUserEnvProbe(params, containerProperties, config, 'cat /proc/self/environ', '\0');
+	if (!env) {
+		params.output.write('userEnvProbe: falling back to printenv');
+		env = await runUserEnvProbe(params, containerProperties, config, 'printenv', '\n');
+	}
+
+	if (env) {
+		await updateUserEnvCache(env, params.containerSessionDataFolder, containerProperties.shellServer);
+	}
+
+	return env || {};
+}
+
+async function readUserEnvFromCache(cacheDir?: string, shellServer?: ShellServer) {
+	if (!shellServer || !cacheDir) {
+		return undefined;
+	}
+
+	const cacheFile = path.posix.join(cacheDir, 'env.json');
+	try {
+		if (await isFile(shellServer, cacheFile)) {
+			const { stdout } = await shellServer.exec(`cat '${cacheFile}'`);
+			return JSON.parse(stdout);
+		}
+	}
+	catch (e) {
+
```

Comment:
Should we log this?

Reply:
I think that is a good idea. Th `output` object was not available in this method to emit outputs. But I can certainly check what is needed to get it there.

Reply:
Sounds good!

Reply:
done